crates/flux-refineck/src/checker.rs

```diff
@@ -907,6 +907,43 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
         }
     }
 
+    fn check_unsized_cast(&self, _env: &mut TypeEnv, from: &Ty, to: &rustc::ty::Ty) -> Result<Ty> {
+        if let TyKind::Indexed(BaseTy::Ref(_, from_ty, from_mut), _) = from.kind()
+            && let TyKind::Indexed(src_base_ty, _) = from_ty.kind()
+            && let rustc::ty::TyKind::Ref(dst_re, to_ty, to_mut) = to.kind()
+        {
+            // &mut [T; n] -> &mut [T][n] and &[T; n] -> &[T][n]
+            if let rustc::ty::TyKind::Slice(_) = to_ty.kind()
+                && let BaseTy::Array(src_arr_ty, src_n) = src_base_ty
+                && from_mut == to_mut
+            {
+                let idx = Expr::from_const(self.genv.tcx(), src_n);
+                let dst_slice = Ty::indexed(BaseTy::Slice(src_arr_ty.clone()), idx);
+                Ok(Ty::mk_ref(*dst_re, dst_slice, *to_mut))
+            } else
+            // &T  -> & dyn U
+            if let rustc::ty::TyKind::Dynamic(_, _) = to_ty.kind() {
+                self.genv
+                    .refine_default(&self.generics, to)
+                    .with_span(self.body.span())
+            } else {
+                tracked_span_bug!("unsupported Unsize cast: from {from:?} to {to:?}")
+            }
+        } else {
+            // TODO: may be too "conservative"? Can it be unsound?
```

Comment:
@ranjitjhala this is unsound for a `&mut [T;n] -> &mut [T][n]` coercion because one should be *invariant* on `T`. Consider

```rust
#[flux::sig(fn([u8{v: v > 0}; 4]) -> [u8{v: v > 0}; 4])]
pub fn test01(mut b: [u8; 4]) -> [u8; 4] {
    test02(&mut b);
    b
}

// This function is crashing for an unrelated reason but it should verify
#[flux::trusted] 
pub fn test02(x: &mut [u8]) {
    if x.len() > 0 {
        x[0] = 0;
    }
}
```

Reply:
I'm working on a proper fix

Reply:
So should `block_with` be used when the target is a `mut` as done when checking function params in `check_fn_call`?

Reply:
yes, we need to convert the (strong) pointer to a mut reference by blocking the pointee. After that the cast is the same as casting a mut ref. I'll implement that, but before I'm abstracting some of the pointer to ref logic, because we reimplement it in a couple of places and it is a bit subtle.

Reply:
OK so I'll let you deal with this one then?

Reply:
Let me deal with it.